tools/mc-java-protoc/src/main/java/io/spine/tools/mc/java/protoc/field/FieldGen.java

```diff
@@ -73,27 +71,61 @@ private FieldGen(ImmutableList<CodeGenerationTask> tasks) {
     /**
      * Creates a new instance based on the passed Protoc config.
      */
-    @SuppressWarnings("MethodWithMultipleLoops") // Required to configure code generation tasks.
     public static FieldGen instance(SpineProtocConfig spineProtocConfig) {
         checkNotNull(spineProtocConfig);
-        AddFields config = spineProtocConfig.getAddFields();
-
         ImmutableList.Builder<CodeGenerationTask> tasks = ImmutableList.builder();
-        EntityStateConfig entityStateConfig = config.getEntityStateConfig();
-        if (isNotDefault(entityStateConfig)) {
-            tasks.add(new GenerateEntityStateFields(entityStateConfig, factory));
+
+        if (spineProtocConfig.hasEntities()) {
+            ForEntities entities = spineProtocConfig.getEntities();
+            GenerateFields fields = entities.getGenerateFields();
+            if (fields.hasSuperclass()) {
+                tasks.add(new GenerateEntityStateFields(entities, factory));
+            }
+        }
+        if (spineProtocConfig.hasCommands()) {
+            ForSignals signals = spineProtocConfig.getCommands();
+            tasks.addAll(tasksFor(signals));
+        }
+        if (spineProtocConfig.hasEvents()) {
+            ForSignals signals = spineProtocConfig.getEvents();
+            tasks.addAll(tasksFor(signals));
         }
-        for (ConfigByPattern byPattern : config.getConfigByPatternList()) {
-            tasks.add(new GenerateFieldsByPattern(byPattern, factory));
+        if (spineProtocConfig.hasRejections()) {
+            ForSignals signals = spineProtocConfig.getRejections();
+            tasks.addAll(tasksFor(signals));
         }
-        for (ConfigByType byType : config.getConfigByTypeList()) {
-            tasks.add(new GenerateFieldsByType(byType, factory));
+        for (ForMessages group : spineProtocConfig.getMessagesList()) {
+            taskFor(group).ifPresent(tasks::add);
         }
         return new FieldGen(tasks.build());
     }
 
+    private static ImmutableList<GenerateFieldsByPattern> tasksFor(ForSignals forSignals) {
+        GenerateFields generateFields = forSignals.getGenerateFields();
+        if (!generateFields.hasSuperclass()) {
+            return ImmutableList.of();
+        }
+        return forSignals.getPatternList()
+                         .stream()
+                         .map(pattern -> new GenerateFieldsByPattern(
+                                 generateFields.getSuperclass(), pattern, factory
+                         )).collect(toImmutableList());
+    }
+
+    private static Optional<GenerateFieldsByPattern> taskFor(ForMessages forMessages) {
+        GenerateFields generateFields = forMessages.getGenerateFields();
+        if (!generateFields.hasSuperclass()) {
+            return Optional.empty();
+        }
+        Pattern pattern = forMessages.getPattern();
+        GenerateFieldsByPattern task = new GenerateFieldsByPattern(
```

Comment:
We have this `new GeneratedFieldsByPattern()` thing twice in this file. Please consider extracting a method which makes it easier to read and understand.

Reply:
It's a trick. Those are two different overloads of the constructor. One accepts `FilePattern` and the other — `Pattern`.

I've changed the naming a bit so that it's more obvious.